cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -427,7 +433,51 @@ template <>
 hash_value_type __device__ inline SparkMurmurHash3_32<numeric::decimal128>::operator()(
   numeric::decimal128 const& key) const
 {
-  return this->compute<__int128_t>(key.value());
+  // Generates the Spark MurmurHash3 hash value, mimicking the conversion:
+  // java.math.BigDecimal.valueOf(unscaled_value, _scale).unscaledValue().toByteArray()
+  // https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L381
+  __int128_t const val     = key.value();
+  cudf::size_type key_size = sizeof(__int128_t);
+  std::byte const* data    = reinterpret_cast<std::byte const*>(&val);
+
+  // Extract the first bit of the key, which holds the sign.
+  std::byte const sign_bit = data[key_size - 1] & static_cast<std::byte>(0x80);
```

Comment:
But really we should do something like `auto const is_negative = (key < 0);` rather than do bitwise gymnastics to determine the sign. We need the bool once to determine `zero_value` and then we only use the byte representation of `sign_bit` once at the end, which could be replaced by `is_negative ? std::byte{0x80} : std::byte{0x00}`... will think some more.